backend/src/apiserver/resource/resource_manager_test.go

```diff
@@ -208,6 +219,19 @@ func createPipeline(name string) *model.Pipeline {
 		}}
 }
 
+func createRunExpectedWorkflow(params []v1alpha1.Parameter, labels map[string]string, annotations map[string]string, serviceAccount string) *v1alpha1.Workflow {
```

Comment:
How about this @Bobgy should make tests a bit cleaner? We still do the comparison with the metadata but as I mentioned [here](https://github.com/kubeflow/pipelines/pull/3965#discussion_r643325843) I think it might be easier for the reader if it is included in this case.

Reply:
Do you think this is better than the original `addMetadata` version? I feel like that's simpler, because it sticks to workflow struct in interface.

I only had concerns regarding naming. Current implementation is good in a few minor points too, so overall I think the following are good:
1. The method returns a new workflow via deepcopy (so that the method does not mutate its inputs)
2. The method name makes tests readable -- we can tell it adds some metadata that is present on runtime workflows.

WDYT?

Reply:
I feel sorry you had to change the entire test file multiple times, I think we can save some intermediate efforts by commenting ideas and discuss in comments, then finish the code when we agree with each other.

Reply:
I rewrote to use `createRunExpectedWorkflow` since all the tests related to `CreateRun` have the same similar set up required where the `metadata` is a subset only. I don't have any strong feelings for either of the solutions, but maybe bringing back the `addMetadata` is easiest and then have a separate look on the tests to improve the readability?

Reply:
Problem with the current `createRunExpectedWorkflow` interface is that, there are a fixed number of arguments.
Whenever any test calls it, it has to supply all the arguments. However, if we will need to add any new arguments in the future, we'll have to update every test that calls `createRunExpectedWorkflow`.

Also, it creates another layer of abstraction over directly creating workflow structs, but the new abstraction doesn't really provide much value, it only adds metadata by default.

Compared to this, the original `addMetadata` is a lot simpler. I was only asking for a more descriptive name.

Therefore, my proposal would be:

```go
func withCreateRunMetadata(*v1alpha1.Workflow workflow) *v1alpha1.Workflow {
workflow.DeepCopy()
...

(you can adjust the name if you have better ideas)
}
```

Reply:
@NikeNano see above

Reply:
Cool, see your point that it is a simpler interface, will update.